scenario_gym/viewer/opencv.py

```diff
@@ -191,109 +181,229 @@ def render(self, state: State) -> Optional[int]:
+
+        if "text" in self.render_layers:
+            self.render_text(state)
+
+    def get_center_pose(self, state: State, e_ref: str) -> np.ndarray:
+        """Get the pose for the center of the frame."""
+        ego_pose = None
+        if e_ref is not None:
+            entity = state.scenario.entity_by_name(e_ref)
+            if entity is None:
+                entity = state.scenario.entities[0]
+                ego_pose = entity.pose
             else:
-                ego_pose = state.scenario.entities[0].pose
-        else:
+                ego_pose = entity.pose
+        if ego_pose is None:
             ego_pose = np.zeros(6)
-        self.draw_road_network_layers(state.scenario.road_network, ego_pose)
+        return ego_pose
 
+    def get_entity_color(self, entity_idx: int, entity: Entity) -> Color:
+        """Get the color to draw the given entity."""
         if self._entity_colour_dict is None:
-            self._entity_colour_dict = {
-                i: (0, 0, 0)
-                if i == 0
-                else tuple(map(int, np.random.randint(256, size=3)))
-                for i in range(len(state.scenario.entities))
-            }
+            self._entity_colour_dict = {}
+        if entity_idx not in self._entity_colour_dict:
+            if entity_idx == 0:
+                c = (0, 0, 0)
+            else:
+                c = tuple(map(int, np.random.randint(256, size=3)))
+            self._entity_colour_dict[entity_idx] = c
+        return self._entity_colour_dict[entity_idx]
+
+    def draw_entity(
+        self, entity_idx: int, entity: Entity, ego_pose: np.ndarray
+    ) -> None:
+        """Draw the given entity onto the frame."""
+        c = self.get_entity_color(entity_idx, entity)
+
+        # get the bounding box in the global frame
+        bbox_coords = entity.get_bounding_box_points()
+
+        # convert to the egos frame (rotated 90)
+        bbox_in_ego = to_ego_frame(bbox_coords, ego_pose, vertical=True)
+
+        # render front of the entity
+        l = np.linalg.norm(bbox_in_ego[0] - bbox_in_ego[1])
+        w = 0.25 / l
+        front_bbox = np.array(
+            [
+                # 0.5 * (bbox_in_ego[2] + bbox_in_ego[1]),
+                bbox_in_ego[2],
+                bbox_in_ego[1],
+                bbox_in_ego[1] + w * (bbox_in_ego[0] - bbox_in_ego[1]),
+                bbox_in_ego[2] + w * (bbox_in_ego[3] - bbox_in_ego[2]),
+            ]
+        )
+
+        # add to frame
+        xy = vec2pix(bbox_in_ego, mag=self.mag, h=self.h, w=self.w)
+        xy_front = vec2pix(front_bbox, mag=self.mag, h=self.h, w=self.w)
+        cv2.fillPoly(self._frame, [xy], c)
+        cv2.fillPoly(self._frame, [xy_front], self.entity_front_color)
+
+    def render_text(self, state: State) -> None:
+        """Add text to the frame."""
+        v = np.linalg.norm(state.scenario.entities[0].velocity[:3])
+        cv2.putText(
+            self._frame,
+            "Ego speed: {:.4f}".format(v),
```

Comment:
2?

Reply:
sorry where does the 2 fit in here? the rounding?